### PR TITLE
Annotations watcher

### PR DESCRIPTION
**Pull request by thebsdbox**
This now offers the `--annotations <provider>` flag, this is used to watch for BGP annotations being applied and will watch for: 

`<provider>/src-ip`
`<provider>/node-asn`
`<provider>/peer-ip`
`<provider>/peer-asn`

Once all these have been populated the watch ends and starts the BGP server

### CODE REVIEW

**Comment by deitch** [comment]
A few questions:

* Are the annotations configurable? 
* Are the docs updated to reflect this option?
* Will the whole "packet" option go away, as this can just use those?

**Review by gianarb — APPROVED** [review]
As we discussed in issue #153 I understand the feature and I know why it is coded in this way (with the possibility to override the prefix). I can't buy that flexibility. It looks a bit confusing for somebody that does not have the same context we have (and the opportunity to ping Dan on a daily bases).

But code is code and it implemented a needed feature!

**Comment by thebsdbox** [comment]
> As we discussed in issue #153 I understand the feature and I know why it is coded in this way (with the possibility to override the prefix). I can't buy that flexibility. It looks a bit confusing for somebody that does not have the same context we have (and the opportunity to ping Dan on a daily bases).
> 
> But code is code and it implemented a needed feature!

If there is a better alternative i'd be happy to implement it? I just need some standard for getting BGP configuration from annotations?

**Comment by gianarb** [comment]
> If there is a better alternative i'd be happy to implement it? I just need some standard for getting BGP configuration from annotations?

I think you should not require a provider as a prefix, you can make your own one. It can be `kube-vip` or `bgp-configure`. It will make kube-vip easier but it will probably require CCM to adapt to your standard (but I think it is something CCM already has to do because only Packet CCM will work out of the box).
If I am wrong all the CCM already set BGP in the right way what I am saying does not have any sense!

**Comment by detiber** [comment]
> > If there is a better alternative i'd be happy to implement it? I just need some standard for getting BGP configuration from annotations?
> 
> I think you should not require a provider as a prefix, you can make your own one. It can be `kube-vip` or `bgp-configure`. It will make kube-vip easier but it will probably require CCM to adapt to your standard (but I think it is something CCM already has to do because only Packet CCM will work out of the box).

+1 on a kube-vip specific the prefix here...  I would say that something like:
`bgp.kube-vip.io/<attribute>` would probably be a good convention to follow for bgp-related annotations.

> If I am wrong and all the CCM implementations already set BGP in the right way what I am saying does not have any sense!

I'm not aware of any other CCM implementations that are interacting with any BGP related config outside of ours.

That said, to ease migrations it might make sense to have some backwards compat support on the kube-vip side at least temporarily.

File: pkg/manager/watcher.go [hunk 1]
```diff
@@ -89,3 +92,120 @@ func (sm *Manager) servicesWatcher(ctx context.Context) error {
...
+	}
+
+	rw, err := watchtools.NewRetryWatcher("1", &cache.ListWatch{
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return sm.clientSet.CoreV1().Nodes().Watch(context.Background(), listOptions)
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("error creating annotations watcher: %s", err.Error())
+	}
```
Comment (detiber) [inline pkg/manager/watcher.go:121]: This is likely a good candidate to use a metadata only client/informer: https://godoc.org/k8s.io/client-go/metadata

Comment (thebsdbox) [reply pkg/manager/watcher.go:122]: I'll have to investigate that 🧐

**Pull request merged**